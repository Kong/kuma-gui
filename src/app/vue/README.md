---
section: Packages
---
# vue

Automatic `vue` initialization using our projects dependency injection layer.

