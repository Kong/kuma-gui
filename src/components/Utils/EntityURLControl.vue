<template>
  <div
    v-if="shouldDisplay"
    data-testid="entity-url-control"
  >
    <KClipboardProvider v-slot="{ copyToClipboard }">
      <KPop placement="bottom">
        <KButton
          appearance="secondary"
          size="small"
          @click="() => { copyToClipboard(shareUrl) }"
        >
          <template v-slot:icon>
            <KIcon
              view-box="0 0 16 16"
              icon="externalLink"
            />
          </template>
          {{ copyButtonText }}
        </KButton>
        <template v-slot:content>
          <div>
            <p>{{ confirmationText }}</p>
          </div>
        </template>
      </KPop>
    </KClipboardProvider>
  </div>
</template>

<script >
export default {
  name: 'EntityURLControl',
  props: {
    name: {
      type: String,
      required: true,
    },
    copyButtonText: {
      type: String,
      default: 'Copy URL',
    },
    confirmationText: {
      type: String,
      default: 'URL copied to clipboard!',
    },
  },
  computed: {
    shareUrl() {
      const urlRoot = `${window.location.href.replace(window.location.hash, '')}#`

      return `${urlRoot}${this.$route.fullPath}?ns=${this.name}`
    },
    shouldDisplay() {
      const mesh = this.$route.params.mesh || null

      // we only want to display the copy button when the user has filtered
      // the view by mesh and not all meshes
      if (mesh && mesh !== 'all') {
        return true
      }

      return false
    },
  },
}
</script>
