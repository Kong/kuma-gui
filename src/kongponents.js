import Vue from 'vue'
import '@kongponents/styles/styles.css'
import KAlert from '@kongponents/kalert'
import KBadge from '@kongponents/kbadge'
import KButton from '@kongponents/kbutton'
import KCard from '@kongponents/kcard'
import KEmptyState from '@kongponents/kemptystate'
import KIcon from '@kongponents/kicon'
import KTable from '@kongponents/ktable'
import KToaster from '@kongponents/ktoaster'
import KoolTip from '@kongponents/kooltip'
import KClipboardProvider from '@kongponents/kclipboardprovider'
import Krumbs from '@kongponents/krumbs'
import KPop from '@kongponents/kpop'
import KToggle from '@kongponents/ktoggle'
import KTabs from '@kongponents/ktabs'
import KModal from '@kongponents/kmodal'
import KInput from '@kongponents/kinput'
import KInputSwitch from '@kongponents/kinputswitch'

Vue.component('KAlert', KAlert)
Vue.component('KBadge', KBadge)
Vue.component('KButton', KButton)
Vue.component('KCard', KCard)
Vue.component('KEmptyState', KEmptyState)
Vue.component('KIcon', KIcon)
Vue.component('KTable', KTable)
Vue.component('KToaster', KToaster)
Vue.component('KoolTip', KoolTip)
Vue.component('KClipboardProvider', KClipboardProvider)
Vue.component('Krumbs', Krumbs)
Vue.component('KPop', KPop)
Vue.component('KTabs', KTabs)
Vue.component('KToggle', KToggle)
Vue.component('KInput', KInput)
Vue.component('KInputSwitch', KInputSwitch)
Vue.component('KModal', KModal)
