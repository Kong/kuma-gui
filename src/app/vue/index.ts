import { createApp } from 'vue'
import {
  createRouter,
  createWebHistory,
} from 'vue-router'

import type { ServiceDefinition } from '@/services/utils'
import { token, createInjections } from '@/services/utils'
import type { Component } from 'vue'
import type { Router, RouteRecordRaw, NavigationGuard } from 'vue-router'

type Token = ReturnType<typeof token>

type VueApp = ReturnType<typeof createApp>
export type PluginDefinition = Parameters<VueApp['use']>
export type ComponentDefinition = [string, Component]

const $ = {
  app: token<(App: Component) => Promise<VueApp>>('vue.app'),
  router: token<Router>('vue.router'),

  components: token('vue.components'),
  plugins: token<PluginDefinition[]>('vue.plugins'),
  routes: token<RouteRecordRaw[]>('vue.routes'),
  routesLabel: token<RouteRecordRaw[]>('vue.routes.label'),
  navigationGuards: token<NavigationGuard[]>('vue.routes.navigation.guards'),
}
export const services = (app: Record<string, Token>): ServiceDefinition[] => {
  return [
    [$.app, {
      service: (
        components: ComponentDefinition[],
        plugins: PluginDefinition[],
      ) => {
        return async (App: Component) => {
          const app = createApp(App)

          plugins.forEach(([...args]) => {
            app.use(...args)
          })

          components.forEach(([name, component]: [string, Component]) => {
            app.component(name, component)
          })

          return app
        }
      },
      arguments: [
        $.components,
        $.plugins,
      ],
    }],
    [$.router, {
      service: (
        env: (str: 'KUMA_BASE_PATH') => string,
        routes: RouteRecordRaw[],
        guards: NavigationGuard[],
      ) => {
        const router = createRouter({
          history: createWebHistory(env('KUMA_BASE_PATH')),
          routes,
        })

        guards.forEach((item) => {
          if (typeof item === 'function') {
            router.beforeEach(item)
          }
        })

        return router
      },
      arguments: [
        app.env,
        $.routes,
        $.navigationGuards,
      ],
    }],

    [token('vue.plugins'), {
      service: (router) => {
        return [
          [router],
        ]
      },
      arguments: [
        $.router,
      ],
      labels: [
        $.plugins,
      ],
    }],

    // TODO(jc): Label decoration. Make it so we can decorate labels
    // Temporarily turn the routesLabel label into routes for decoration
    // purposes
    [$.routes, {
      service: (routes: RouteRecordRaw[]) => routes,
      arguments: [
        $.routesLabel,
      ],
    }],
    //

    // TODO(jc): Make it so we don't need to provide a [] empty default for
    // labels
    [token('application.routes.navigation.guards'), {
      service: () => {
        return []
      },
      labels: [
        $.navigationGuards,
      ],
    }],
  ]
}

export const TOKENS = $
export const [
  useRouter,
] = createInjections(
  TOKENS.router,
)
