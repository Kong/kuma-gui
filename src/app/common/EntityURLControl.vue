<template>
  <div data-testid="entity-url-control">
    <KClipboardProvider v-slot="{ copyToClipboard }">
      <KPop placement="bottom">
        <KButton
          appearance="outline"
          icon="copy"
          size="small"
          @click="() => { copyToClipboard(shareUrl) }"
        >
          Copy URL
        </KButton>

        <template #content>
          <p>URL copied to clipboard!</p>
        </template>
      </KPop>
    </KClipboardProvider>
  </div>
</template>

<script lang="ts" setup>
import { computed } from 'vue'
import { useRoute, useRouter } from 'vue-router'
import { KClipboardProvider, KPop, KButton } from '@kong/kongponents'

const route = useRoute()
const router = useRouter()

const props = defineProps({
  name: {
    type: String,
    required: false,
    default: '',
  },

  mesh: {
    type: String,
    required: false,
    default: '',
  },

  route: {
    type: Object,
    required: false,
    default: null,
  },
})

const shareUrl = computed(() => {
  const urlRoot = `${window.location.href.replace(window.location.hash, '')}#`
  const shareRoute = props.route !== null
    ? props.route
    : {
      name: route.name,
      params: { mesh: props.mesh },
      query: { ns: props.name },
    }
  const resolvedRoute = router.resolve(shareRoute)

  return `${urlRoot}${resolvedRoute.fullPath}`
})
</script>
