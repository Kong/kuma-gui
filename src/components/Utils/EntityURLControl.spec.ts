import { render, screen } from '@testing-library/vue'
import { KButton, KClipboardProvider, KIcon, KPop } from '@kong/kongponents'

import EntityURLControl from './EntityURLControl.vue'

function renderComponent(props: any, mesh = 'default') {
  return render(EntityURLControl, {
    global: {
      components: {
        KButton,
        KClipboardProvider,
        KIcon,
        KPop,
      },
      mocks: {
        $route: {
          params: {
            mesh,
          },
        },
      },
    },
    props,
  })
}

describe('EntityURLControl.vue', () => {
  it('renders snapshot', () => {
    const { container } = renderComponent({
      name: 'foo',
    })

    expect(container).toMatchSnapshot()
  })

  it('render for mesh all', () => {
    renderComponent({
      name: 'foo',
    }, 'all')

    expect(screen.queryByTestId('entity-url-control')).toBeInTheDocument()
  })
})
