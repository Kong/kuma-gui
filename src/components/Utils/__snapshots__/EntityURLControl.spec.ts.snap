// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EntityURLControl.vue renders snapshot 1`] = `
<div>
  <div
    data-testid="entity-url-control"
  >
    <div
      data-v-3cb990b6=""
    >
      <button
        class="k-button small secondary"
        data-v-38fa8ecf=""
        data-v-3cb990b6=""
        type="button"
      >
        <svg
          class="kong-icon kong-icon-externalLink"
          data-v-38fa8ecf=""
          data-v-a7c9c1d2=""
          height="12"
          role="img"
          viewBox="0 0 16 16"
          width="12"
        >
          <title
            data-v-a7c9c1d2=""
          >
            externalLink
          </title>
          <g
            data-v-a7c9c1d2=""
          >
            <path
              d="M8.24308 1.5C7.83269 1.5 7.5 1.16710138 7.5.75c0-.41421356.33823-.75.74308-.75h3.00568C11.6666 0 12 .336341 12 .75123866v3.00567786C12 4.1673102 11.6671 4.5 11.25 4.5c-.41421 0-.75-.3382314-.75-.74308348V2.625L6.53246 6.59254c-.28982.2898217-.76955.2953689-1.06202.0029022l-.06588-.0658844c-.29098-.2909742-.29117-.7679483.0029-1.0620178L9.375 1.5H8.24308zm-6.74753 0C.67089 1.5 0 2.1695784 0 2.99554521v7.50890959C0 11.3291147.66958 12 1.49555 12h7.5089C9.82911 12 10.5 11.3304216 10.5 10.5044548V6.75H9v3.75H1.5V3h3.75V1.5H1.49555z"
              data-v-a7c9c1d2=""
              fill="#000"
              fill-opacity=".25"
              fill-rule="evenodd"
            />
          </g>
        </svg>
        
        Copy URL
      
      </button>
      <transition-stub
        data-v-3cb990b6=""
        name="fade"
      >
        <div
          class="k-popover"
          data-v-3cb990b6=""
          style="width: 200px; display: none;"
        >
          <!---->
          <div
            class="k-popover-content"
            data-v-3cb990b6=""
          >
            <div>
              <p>
                URL copied to clipboard!
              </p>
            </div>
          </div>
        </div>
      </transition-stub>
    </div>
  </div>
</div>
`;
